### PR TITLE
Extend nilChecking to account for if-statements

### LINKED ISSUE #11918: nil checker catches programs that should be valid
The nil checker should probably know not to look into conditionals that can't be reached. The programs below are safe and reasonable to write during development.

``` chapel
class C { var x: int; }

proc main() {
  var obj: unmanaged C?;

  if obj then
    writeln(obj.x);
}

```

``` chapel
proc f() {
  return 1;
}

proc main() {
  try {
    f();
  } catch e: Error {
    writeln(e.message());
  }
}
```

Tests:
```
# the first example above
test/classes/vass/if-object-2.chpl
# see also
test/classes/nilability/if-object-1.chpl
# the second example above
test/classes/nilability/catch-no-throw.chpl
```

Updated 2019-10-04:
* `catch-no-throw.chpl` (the second example above) works now
* `if-object-2.chpl` (the first example above) works when using owned/shared/borrowed instead of unmanaged

Updated 2019-12-10:
* `if-object-2.chpl` (the first example above) now works with unmanaged, too, as of #14579 / b9fdd4ced1, so it is [no longer a future](https://github.com/chapel-lang/chapel/pull/14620).
* Still, it is unlikely that the underlying shortcoming has been fixed. We need either to build a reproducer or else declare this issue resolved.

### PR DESCRIPTION
**Pull request by vasslitvinov**
Resolves #11918. Removes some unnecessary calls to postfix-`!`.
Also addresses #13630, I plan to close it with a separate PR.

The nilChecking pass now understands the effect of conditionals that compare
a class variable against `nil`.

* Within the then/else branches of such a conditional, the variable is known
to start out as nil or non-nil, as appropriate.

* If the branch will never be executed at runtime, nilChecking does not
analyze it and so does not issue "you are dereferencing nil" errors
that it would otherwise.

* Calls to postfix-`!` that are known to succeed or never executed
are removed. This way, they do not incur runtime cost and do not
potentially disable some optimizations.

For example:
```chpl
    var x: owned MyClass?;
    var y = ....;

    if x != nil then
      writeln(y!);     // never executed, an error is never issued

    if y != nil then
      writeln(y!);     // the runtime ! check is removed
```
This does not address field accesses, partly because it is hard for
the compiler to ensure that the field is not modified concurrently.

#### Next steps

* Analyze conditionals like these:
```chpl
    var x = ...;
    if x == nil then
      x = new MyClass();  // (a)
      return;             // (b)
      exit();             // (c)

    writeln(x!);          // x is never nil if (a), (b), or (c) happens
```

* Potentially revisit the implementation choices - see below.

#### Implementation Notes

I considered these approaches to handling unreachable branches:

* Remove them from the AST altogether.
* Do not analyze; remove postfix-! calls from them.
* Analyze normally, as if they are reachable.

I chose the second option as a middle ground. The first option is aggressive
and I did not have enough confidence in the correctness of nilChecking.
The situation when a branch is known to be unreachable due to a nil check
is a corner case, which is hard to test thoroughly.

The third option is the most conservative. I chose against it because
it comes with this challenge:
```chpl
    var x = nil;

    if x != nil then
      // in the then-branch the compiler knows x!=nil
    else
      // here the compiler knows x == nil

    // To the compiler, since x!=nil in one branch and x==nil in the other branch,
    // after the conditional x can be either after. So x==nil is no longer known.
```

Another dimension is how to handle variables, esp. formals, of non-nilable types.
The could start out either as "allocated" or as "unknown". I chose "unknown"
as the more conservative option because above I chose to allow the analysis
to remove `!` checks and to suppress nil-access errors. The challenge with
this choice comes from the possibility for a non-nilable variable to contain
`nil`, due to holes in our compile-time checking of nilable types.
We can change all these choices as we gain more experience.

The nilChecking code I added propagates the "is non/nil" information up the
use-def chains through casts. This is because, prior to testing `if x != nil`,
the compiler casts `x` to `object` if it is not already.

While there, ensure that formals are initially "unknown". Also, make `combine()`
insensitive w.r.t. an AliasLocation's `location`, except when the `type` is REF.
I.e. when combining two AliasLocations of the same type, produce the same
type as a result even if the two locations differ. This is because the location
is used merely for line number reporting. REF is an exception because
the referent matters for the analysis, not just for the line number.

Minor: I slightly tuned up the debugging output.

Minor: gatherVariablesToCheck() right now serves only to detect the
functions that have no class variables to be analyzed. We can simplify it
to do only that.

Test changes: I combined the two versions of `if-object-2` into one
and moved it next to `if-object-1.chpl`. In a couple of other tests
a runtime failure of a `!`-check is replaced with a compile-time error.

Testing: linux64 -futures, multilocale tests under gasnet.

### CODE REVIEW

**Comment by mppf** [comment]
> Once we eliminate holes in our compile-time checking of nilable types,
we can also remove ! on formals of non-nilable types.

The implementation of these in ChapelBase does not add a check. Instead, for non-nilable types, we are relying on the old checking upon a method call. So, this PR could remove such checks. (But it should not matter).

File: compiler/resolution/nilChecking.cpp [hunk 1]
```diff
@@ -663,27 +848,43 @@ static void adjustMapForCatchBlock(CondStmt* cond, bool inThenBranch,
...
+        updateForConditional(errorSym, inThenBranch, cond->condExpr, OUT);
       }
     }
   }
 }
 
 // Update 'OUT' based on being inside a conditional,
 // if 'bb' is the starting BasicBlock of the then- or else- branch.
-static void adjustMapForConditional(BasicBlock* bb, AliasMap& OUT) {
+static bool adjustMapForConditional(BasicBlock* bb, AliasMap& OUT,
```
Comment (mppf) [inline compiler/resolution/nilChecking.cpp]: Add a comment saying it returns true if the block is unreachable

File: compiler/resolution/nilChecking.cpp [hunk 2]
```diff
@@ -782,6 +993,23 @@ static void gatherVariablesToCheck(FnSymbol* fn,
   }
 }
 
+// The formals are initially unknown.
+// When we are confindent in our nilable type-checking, set the formals
+// whose types are non-nilable to MUST_ALIAS_ALLOCATED.
```
Comment (mppf) [inline compiler/resolution/nilChecking.cpp]: I think this should set formals of non-nilable class type to MUST_ALIAS_ALLOCATED now - see my earlier comment on the PR

File: compiler/resolution/nilChecking.cpp [hunk 3]
```diff
@@ -644,6 +724,111 @@ static Expr* getSingleDefExpr(Expr* expr) {
...
+    // None of the above cases - do not continue.
+    break;
+  }
+  return false;
+}
+
+// Is 'expr' a SymExpr for 'nil' ?
+// I consider it insufficient to test "expr->getValType() == dtNil"
+// because "expr" could be a calltemp where the invoked function
+// side-effects the variable involved in the comparison, as a corner case.
```
Comment (mppf) [inline compiler/resolution/nilChecking.cpp]: I would have thought checking against dtNil would be OK, and I don't follow your rationale here. Could you share an example?

File: compiler/resolution/nilChecking.cpp [hunk 4]
```diff
@@ -644,6 +724,111 @@ static Expr* getSingleDefExpr(Expr* expr) {
...
+    if (it != OUT.end()) {
+      AliasType altype = it->second.type;
+      if ((isAllocated && altype == MUST_ALIAS_NIL)        ||
+          (!isAllocated && altype == MUST_ALIAS_ALLOCATED) )
+        // We are testing ==nil on something that's not nil, or visa versa.
+        return true;
+    }
+
+    // Not worth updating OUT if 'curr' is the only use (a common case?).
+    if (curr->symbol()->getSingleUse() == curr) ;
```
Comment (mppf) [inline compiler/resolution/nilChecking.cpp]: Wouldn't it be better to write it as `if (curr->symbol()->getSingleUse() != curr) updateForConditional(...)` ?

Does this actually matter for performance? For correctness?

File: compiler/resolution/nilChecking.cpp [hunk 5]
```diff
@@ -644,6 +724,111 @@ static Expr* getSingleDefExpr(Expr* expr) {
   return NULL;
 }
 
+// SE->symbol() is known to be nil/non-nil. If it comes from a MOVE or CAST,
+// propagate to their arguments.
+// For efficiency, the caller should assure isSymbolAnalyzed(SE).
+// Returns true and skips updating 'OUT' if the branch in question is unreachable.
+static bool adjustTestArgChain(SymExpr* SE, bool isAllocated, AliasMap& OUT) {
+  SymExpr* curr = SE;
+  while (true) {
```
Comment (mppf) [inline compiler/resolution/nilChecking.cpp]: I'd appreciate a comment here saying something like "loop over the definitions for temporary symbols"

File: compiler/resolution/nilChecking.cpp [hunk 6]
```diff
@@ -644,6 +724,111 @@ static Expr* getSingleDefExpr(Expr* expr) {
...
+    if (curr->symbol()->getSingleUse() == curr) ;
+    else updateForConditional(curr->symbol(), isAllocated, curr, OUT);
+
+    if (! curr->symbol()->hasFlag(FLAG_TEMP))
+      // Stop following the use->def chain to safeguard against corner cases
+      // of user var being concurrently modified after being read into 'curr'.
+      break;
+
+    if (CallExpr* CE = toCallExpr(getSingleDefExpr(curr)))
+      if (CE->isPrimitive(PRIM_MOVE) || CE->isPrimitive(PRIM_CAST))
```
Comment (mppf) [inline compiler/resolution/nilChecking.cpp]: I think it's good practice to also check for PRIM_ASSIGN anywhere you check PRIM_MOVE (unless the difference between move/assign for references is relevent - and I don't think it is here)

File: compiler/resolution/nilChecking.cpp (hunk 2 above)
Comment (vasslitvinov) [reply compiler/resolution/nilChecking.cpp]: I added a discussion of this to the PR message and kept "unknown" for now.

**Pull request merged**